### ISSUE TITLE
Client-Side Validation not working

### ISSUE THREAD

**Issue by tkardaridis** [JeremySkinner/FluentValidation#960]
### System Details

- FluentValidation version: 8.0.100
- Web Framework version (eg ASP.NET Core 2.1, MVC 5, WebApi 2. Delete if not applicable): MVC 5 

### Issue Description
I have a partial view which is loaded through ajax. The partial view has an Ajax.BeginForm as follows:

@using (Ajax.BeginForm("SubmitTravelStoryAjax", "TravelStories", null, new AjaxOptions
                        {
                            HttpMethod = "POST",
                            InsertionMode = InsertionMode.Replace,
                            OnFailure = "ShowErrorOnPostTravelStory",
                            OnSuccess = "ShowSuccessOnPostTravelStory",
                            LoadingElementId = "divPreLoader"
                        },
                        new
                        {
                            id = "formPostStory",
                            enctype = "multipart/form-data"
                        }))
                        {
                            @Html.ValidationSummary(true)

                            @Html.AntiForgeryToken()
                            @Html.HiddenFor(model => model.ContentId)

                            <div class="form-group form-group-icon-left">
                                <label>@umbraco.library.GetDictionaryItem("Members.TravelStories.NewStoryTitle")</label>
                                <div class="tabbable">
                                    <ul class="nav nav-tabs" id="myTab">
                                        @foreach (var story in Model.StoryTitleCollection)
                                        {
                                            <li class="@(story.CultureCode.ToLower() == System.Threading.Thread.CurrentThread.CurrentUICulture.TwoLetterISOLanguageName.ToLower() ? "in active" : "")">
                                                <a href="#tab-title-@story.CultureCode" data-toggle="tab">@story.CultureCode</a>
                                            </li>
                                        }
                                    </ul>
                                    <div class="tab-content" id="tab-content">
                                        @Html.EditorFor(model => model.StoryTitleCollection, new { htmlAttributes = new { @class = "form-control" } })
                                    </div>
                                </div>
                            </div>
                            
                            <div class="form-group">
                                <input class="btn btn-primary btn-fullwidth" value="@umbraco.library.GetDictionaryItem("ButtonSave")" type="submit" id="btnSaveStory" />
                            </div>
                        }

As you can see within the form is a <ul> tabs with two tabs, one for each language supported by the site and a <div> TabContent which contains the Editor for the List<StoryTitle> class. The code for the editor is the following:

@model Umbraco.Models.SurfaceModels.Content.StoryTitle

<div class="tab-pane fade pt10 pb2 pl2 pr2 @(@Model.CultureCode.ToLower() == System.Threading.Thread.CurrentThread.CurrentUICulture.TwoLetterISOLanguageName.ToLower() ? "in active" : "")" id="tab-title-@Model.CultureCode">
    @Html.HiddenFor(model => model.CultureCode)
    @Html.TextBoxFor(model => model.Title, new { @placeholder = umbraco.library.GetDictionaryItem("Members.TravelStories.NewStoryTitle"), @class = "form-control" })
    @Html.ValidationMessageFor(model => model.Title, "", new { @class = "text-danger" })
</div>

As you can see also inside the editor is a TextBoxFor for the Title property and a ValidationMessageFor for the title also.

I had create a validator for my model as follows:

public class TravelStoryViewModelValidator : AbstractValidator<TravelStoryViewModel>
    {
        public TravelStoryViewModelValidator()
        {
            CascadeMode = CascadeMode.StopOnFirstFailure;
            RuleForEach(list => list.StoryTitleCollection)
                .SetValidator(new TitleIsRequiredValidator())
                .WithLocalizedMessage(typeof(Resources.ClientValidationMessages), "TravelStory_TitleIsRequired");
        }
    }

What i want to do? I want to validate the title as required depending on client language. If the UI culture is EN then i want the EN title to be Required, if the language is EL i want the EL title to be required. Because of that and following a lot documentation from the net i wrote the TitleIsRequiredValidator as follows:

internal interface ITitleIsRequiredValidator { }

    public class TitleIsRequiredValidator : PropertyValidator, ITitleIsRequiredValidator
    {
        public TitleIsRequiredValidator()
            : base("TravelStory_TitleIsRequired", typeof(Resources.ClientValidationMessages))
        {
        }

        protected override bool IsValid(PropertyValidatorContext context)
        {
            if (context.ParentContext.InstanceToValidate.GetType().IsAssignableFrom(typeof(StoryTitle)))
            {
                // Instance of the class which contains property which must be validated 
                var model = context.ParentContext.InstanceToValidate as StoryTitle;
                var storyTitle = context.PropertyValue as string;

                if (model.CultureCode.ToLower() == System.Threading.Thread.CurrentThread.CurrentCulture.TwoLetterISOLanguageName.ToLower())
                {
                    return true;
                }
                else
                {
                    return !string.IsNullOrEmpty(storyTitle);
                }
            }

            return true;
        }
    }

    public class TitleIsRequiredValidatorAdaptor : FluentValidationPropertyValidator
    {
        private const string ValidationType = "TitleRequired";

        public TitleIsRequiredValidatorAdaptor(ModelMetadata metadata, ControllerContext controllerContext, PropertyRule rule, IPropertyValidator validator)
            : base(metadata, controllerContext, rule, validator)
        {
        }

        public override IEnumerable<ModelClientValidationRule> GetClientValidationRules()
        {
            if (!ShouldGenerateClientSideRules()) yield break;

            var validator = Validator as ITitleIsRequiredValidator;

            var formatter = new MessageFormatter().AppendPropertyName(Rule.PropertyName);
            string message = formatter.BuildMessage(Validator.Options.ErrorMessageSource.GetString(null));

            yield return new ModelClientValidationRule
            {
                ValidationType = ValidationType,
                ErrorMessage = message
            };
        }
    }

And because this is a complex validation is getting fired only when the submit button is pressed and validated server side. I need the validation to work also as client side. Can you please help me what else to do in order to enable that feature also?

I also add the validation in global.asax as follows:

FluentValidationModelValidatorProvider.Configure(provider => {
                provider.Add(typeof(Models.Validators.TitleIsRequiredValidator), (metadata, context, rule, validator) => new Models.Validators.TitleIsRequiredValidatorAdaptor(metadata, context, rule, validator));
            });

What else is missing in order to get fired and validate client-side as expected?

Thanks a lot

**Comment by JeremySkinner** [comment]
Hi there. FluentValidation does not provide client-side validation directly. Client-side validation is done by jQuey Validate. Out of the box MVC provides several adapters that generate metadata from serverside rules that can be used by jquery.Validate on the client. FluentValidatoon makes use of these, which is why you get clientside validation for simple rules (such as NotNull/Length etc)

For anything more complex you will need to write the clientside jQuery Logic yourself. Your adapter class needs to return a custom ModeClientValidationRule (this is part of MVC not FluentValidaion) that will generate the metadata attributes that will be inserted into the html element. You will then need to write the appropriate javadcript for jquery.validate to handle these attributes. Note that none of this process is related specifically to FluentValidation, so I can’t really help you further with this - you’ll need to read the MVC documentation on creating custom clientside rules, as well as look at the jquery.validate documentation on how to plug in the custom JavaScript.

**Comment by tkardaridis** [comment]
Thanks for the reply. On this [issue](https://github.com/JeremySkinner/FluentValidation/issues/73) which will probably do my job you are talking about ChildCollectionValidatorAdaptor. Can you please explain or give an example of ChildCollectionValidatorAdaptor.

Thanks

**Comment by JeremySkinner** [comment]
I'm not sure what you're asking...`ChildCollectionValidatorAdapter` is how `RuleForEach` is implemented internally...it's not related to client-side validation at all.

**Comment by tkardaridis** [comment]
Because i tried a lot with javascript coding and it does not work i saw that [issue](https://github.com/JeremySkinner/FluentValidation/issues/73). Probably with the solution you mention in that issue using nestedrules  i could add a rule for that is supported client-side each property of my objects within the list, and it might work. For example

RuleFor(list => list.StoryTitleCollection)
                .NestedRules(v => {
                    v.RuleFor(storyTitle => storyTitle.Title)
                        .NotEmpty()
                        .When(storyTitle => storyTitle.CultureCode.ToLower().Equals(System.Threading.Thread.CurrentThread.CurrentUICulture.TwoLetterISOLanguageName.ToLower()))
                        .WithLocalizedMessage(typeof(Resources.ClientValidationMessages), "TravelStory_TitleIsRequired");
                });

**Comment by JeremySkinner** [comment]
No that won’t help you. 

That rule can not be executed on the clientside - it’s too complex. NestedRules like this can’t be converted to clientside rules, and neither can rules with conditions (using When). This can only be executed on the server. 

Only a small number of rules can be represented on the client side - the documentation is clear on this (see the clientside validation section on https://fluentvalidation.net/aspnet). Nothing custom or with conditions can be automatically converted to client-side rules. 

As I said before, you’ll need to write custom JavaScript for this if you want to validate client side.

**Comment by tkardaridis** [comment]
OK. That is clear.
You were extremely helpful, and to be truly, fluentvalidation is a great project. 
Thanks a lot.
